webui/react/src/components/UPlot/UPlotChart.tsx

```diff
@@ -16,151 +16,239 @@ export interface Options extends Omit<uPlot.Options, 'width'> {
+    () =>
+      getExtendedOptions(
+        options,
+        boundsRef,
+        chartDivRef.current?.offsetWidth,
+        () => setIsReady(true),
+        () => setIsReady(false),
+      ),
+    [ options ],
+  );
+  const previousExtendedOptions = usePrevious(extendedOptions, undefined);
+
+  useEffect(() => {
     return () => {
-      plotChart.destroy();
+      chartRef?.current?.destroy();
       chartRef.current = undefined;
     };
-  }, [ chartDivRef, hasData, normalizedData, options ]);
+  }, []);
 
-  /*
-   * Chart data when data changes.
-   */
   useEffect(() => {
-    if (!chartRef.current || !normalizedData) return;
-    chartRef.current.setData(normalizedData, isZoomed.current);
-  }, [ normalizedData ]);
+    if (!chartDivRef.current) return;
+    if (shouldRecreate(previousExtendedOptions, extendedOptions, chartRef.current)) {
+      /**
+       * TODO: instead of returning true or false,
+       * return a list of actions/payloads to dispatch
+       */
+      chartRef.current?.destroy();
+      chartRef.current = undefined;
 
-  /*
+      const isZoomed = Object.values(boundsRef.current).some((i) => i.isZoomed === true);
+      if (!isZoomed) {
+        boundsRef.current = {};
+      } else {
+        /**
+         * TODO: preserve zoom when new series is selected?
+         * There are some additional challenges because the setDatas will be interpreted as
+         * zooms when the data is streaming in since the bounds are smaller at first
+         * Might also want to preserve other user interactions with the charts
+         * by taking some things from chartRef.current and putting them in newOptions
+         * e.g. a series is updated, say it's hidden, that update is reflected in options
+         * but series is a list, and uPlot.assign does not merge the lists, it clobbers.
+         * (as one would expect)
+         */
+      }
+      try {
+        chartRef.current = new uPlot(extendedOptions, data as AlignedData, chartDivRef.current);
+      } catch(e) {
+        chartRef.current?.destroy();
+        chartRef.current = undefined;
+        handleError(e, {
+          level: ErrorLevel.Error,
+          publicMessage: 'Unable to Load data for chart',
+          publicSubject: 'Bad Data',
+          silent: false,
+          type: ErrorType.Ui,
+        });
+      }
+    } else {
+      const isZoomed: boolean = Object.values(boundsRef.current).some((i) => i.isZoomed === true);
+      const propsProvideScales: boolean = Object.values(extendedOptions?.scales ?? {}).some(
+        (scale) => scale.min != null || scale.max != null,
+      );
+      const onlyOneXValue: boolean = Array.isArray(data) && data[0]?.length === 1;
+      const customScaleIsInEffect = isZoomed || (propsProvideScales && onlyOneXValue);
+      const resetScales = !customScaleIsInEffect;
+
+      try {
+        if (chartRef.current && isReady){
+          chartRef.current.setData(data as AlignedData, resetScales);
+          /**
```

Comment:
so this is where I was initially trying to resolve the issue of the `TrialChart`s not getting updated when switching between trials.

but 
```
chart.setData(data, true)
chart.redraw(true, true)
```
did not result in the `xScale` getting reset (even though it was called for the `yScale`, and even though the x data had different length and bounds).

so instead i did a workaround where I just got the chart to redraw